File: destack/core/common/text.py
from typing import TYPE_CHECKING, Optional

from ..builtin import (
    EnumType,
    FlagEnum,
    Node,
    OptionEnum,
    ReferenceType,
    Struct,
    StructType,
    declare_enum,
    declare_option,
    declare_property,
    declare_struct,
)

if TYPE_CHECKING:
    pass


@declare_enum(EnumType.TEXT_SPAN_TYPE)
class TextSpanType(OptionEnum):
    TEXT = declare_option(1, description="Formatted text")
    HARD_BREAK = declare_option(2, description="Hard break")
    NODE = declare_option(10, description="Reference to a Node")
    LINK = declare_option(11, description="Hyperlink")
    EQUATION = declare_option(20, description="TeX equation")


@declare_enum(EnumType.TEXT_STYLE_FLAG)
class TextStyleFlag(FlagEnum):
    """A flag that can be applied to a TextSpan."""

    DEFAULT = declare_option(0)
    BOLD = declare_option(1)
    ITALIC = declare_option(2)
    STRIKETHROUGH = declare_option(4)
    UNDERLINE = declare_option(8)
    CODE = declare_option(16)


@declare_struct(StructType.TEXT_SPAN)
class TextSpan(Struct):
    """A span of text with optional formatting"""

    type: TextSpanType = declare_property(
        100,
        default=TextSpanType.TEXT,
        tag=None,
    )
    content: Optional[str] = declare_property(
        101,
        tag=None,
    )
    node: Optional[Node] = declare_property(
        102,
        reference_type=ReferenceType.IDENTITY,
        tag=None,
    )
    url: Optional[str] = declare_property(
        105,
        tag=None,
    )
    style_flags: TextStyleFlag = declare_property(
        110,
        tag=None,
        default=TextStyleFlag.DEFAULT,
    )


@declare_struct(StructType.TEXT)
class Text(Struct):
    """
    Rich Text; a single paragraph composed of TextSpans with inline formatting.
    """

    spans: list[TextSpan] = declare_property(
        103,
        tag=None,
    )
    style_flags: TextStyleFlag = declare_property(
        110,
        tag=None,
        default=TextStyleFlag.DEFAULT,
    )

File: destack-rs/destack/src/core/common/text.rs
//! destack.core.common.text@2025.08.15.1

#![destack::partial(destack.core.common.text, file)]

#[destack::generated(Text, -, block)]
/// Text is a single paragraph composed of TextSpans with inline styling.
pub struct Text {
    pub spans: Vec<TextSpan>,
    pub style_flags: TextStyleFlag,
}

#[destack::generated(TextSpan, -, block)]
/// Span of text with optional styling.
pub struct TextSpan {
    pub r#type: TextSpanType,
    pub content: Option<String>,
    pub node: Option<i64 /* TODO */>,
    pub url: Option<String>,
    pub style_flags: TextStyleFlag,
}

#[destack::generated(TextSpanType, -, block)]
/// TextSpanType
pub enum TextSpanType {
    /// Formatted text
    Text = 1,
    /// Hard break
    HardBreak = 2,
    /// Reference to a Node
    Node = 10,
    /// Hyperlink
    Link = 11,
    /// TeX equation
    Equation = 20,
}

#[destack::generated(TextStyleFlag, -, block)]
/// A flag that can be applied to a TextSpan.
pub enum TextStyleFlag {
    Default = 0,
    Bold = 1,
    Italic = 2,
    Strikethrough = 4,
    Underline = 8,
    Code = 16,
}
